src/AudioToolbox/MusicSequence.cs

```diff
@@ -274,6 +288,39 @@ public double GetBeatsForSeconds (double seconds)
 				return beats;
 			return 0;
 		}
+
+		[DllImport (Constants.AudioToolboxLibrary)]
+		extern static /* OSStatus */ MusicPlayerStatus MusicSequenceSetUserCallback (/* MusicSequence */ IntPtr inSequence, MusicSequenceUserCallbackProxy inCallback, /* void * */ IntPtr inClientData);
+
+		public void SetUserCallback (MusicSequenceUserCallback callback)
+		{
+			lock (userCallbackHandles) {
+				if (userCallbackHandles.ContainsKey (handle))
+					userCallbackHandles[handle] = callback;
+				else
+					userCallbackHandles.Add (handle, callback);
+			}
+
+			MusicSequenceSetUserCallback (handle, userCallbackProxy, IntPtr.Zero);
+		}
+
+#if !MONOMAC
+		[MonoPInvokeCallback (typeof (MusicSequenceUserCallbackProxy))]
+#endif
+		static void UserCallbackProxy (IntPtr inClientData, IntPtr inSequence, IntPtr inTrack, double inEventTime, IntPtr inEventData, double inStartSliceBeat, double inEndSliceBeat)
+		{
+			MusicSequenceUserCallback userCallback;
+			lock (userCallbackHandles)
+				userCallbackHandles.TryGetValue (inSequence, out userCallback);
+
+			if (userCallback != null) {
+				var userEventData = new MusicEventUserData (inEventData);
+				var musicSequence = MusicSequence.Lookup (inSequence);
+				var musicTrack = new MusicTrack (musicSequence, inTrack, false);
+
+				userCallback (musicTrack, inEventTime, userEventData, inStartSliceBeat, inEndSliceBeat);
```

Comment:
It doesn't look like the callback gets any big chunks of memory, so I don't think there will be any memory issues. From what I can see these native objects aren't ref counted, and they're not owned by the managed wrapper either, which means native code can free them at any time.